src/smt/theory_str_noodler/formula_preprocess.cpp

```diff
@@ -1520,4 +1530,30 @@ namespace smt::noodler {
         return can_unify(left, right, check);
     }
 
+
+    /**
+     * @brief Adds restrictions from conversions to the len_formula, so that (underapproximating) unsat check can be better
+     * 
+     * Specifically, it checks if for to_code(x)/to_int(x) there is any valid word in the language of automaton for x, i.e,
+     * some one-symbol word for to_code(x) or some word containing only digits for to_int(x).
+     * 
+     * @param conversions 
+     */
+    void FormulaPreprocessor::conversions_validity(std::vector<TermConversion>& conversions) {
+        mata::nfa::Nfa sigma_aut = aut_ass.sigma_automaton();
+        mata::nfa::Nfa only_digits_aut(2, {0}, {1});
+        for (mata::Symbol digit = 48; digit <= 57; ++digit) {
```

Comment:
It would be helpful to extract these magic numbers into named constants.

Reply:
Should  I define them in the function or as some global constants (they are used in multiple places)?

Reply:
I defined them in `AutAssignment`